src/core/iomgr/tcp_server.h

```diff
@@ -80,4 +80,8 @@ int grpc_tcp_listener_get_port(grpc_tcp_listener *listener);
 void grpc_tcp_listener_ref(grpc_tcp_listener *listener);
 void grpc_tcp_listener_unref(grpc_tcp_listener *listener);
 
+/* potentially (trivially) implementable: */
+int grpc_tcp_listener_get_fd(grpc_tcp_listener *listener, unsigned index);
```

Comment:
grpc_tcp_server_get_fd() doesn't index into siblings. I'm not sure that's right.

grpc_tcp_listener has a refcount, but is also a singly-linked list node. What do you do when the refcount of a node in the middle of a singly-linked list drops to zero? This can be OK if the grpc_tcp_listener API never access grpc_tcp_listener::next, because grpc_tcp_server owns refs and is the only API that accesses grpc_tcp_listener::next. I just want to double-check that this is intentionally part of the design.

If grpc_tcp_server_get_fd() can be made to access all the FDs (including siblings), do we need grpc_tcp_listener_get_fd()? on_accept_cb could take the grpc_tcp_server* and an index. I'd also consider moving the refcount into the grpc_tcp_server, so grpc_tcp_listener_{ref,unref}() ref the grpc_tcp_server. For example, in the current scheme, the server can be deleted before the listeners so that grpc_tcp_listener::server points to a deleted object.

It might be worth taking a step back and rethinking some of the design and API goals.